u_roller.go

```diff
@@ -0,0 +1,110 @@
+package tls
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"net"
+	"sync"
+	"time"
+)
+
+type Roller struct {
+	HelloIDs            []ClientHelloID
+	HelloIDMu           sync.Mutex
+	NextHelloID         ClientHelloID
+	TcpDialTimeout      time.Duration
+	TlsHandshakeTimeout time.Duration
+}
+
+// NewRoller creates Roller object with default range of HelloIDs to cycle through until a
+// working/unblocked one is found.
+func NewRoller() (*Roller, error) {
+	tcpDialTimeoutInc, err := getRandUint16()
+	if err != nil {
+		return nil, err
+	}
+	tcpDialTimeoutInc = tcpDialTimeoutInc % 14
+
+	tlsHandshakeTimeoutInc, err := getRandUint16()
+	if err != nil {
+		return nil, err
+	}
+	tlsHandshakeTimeoutInc = tlsHandshakeTimeoutInc % 20
+
+	r := &Roller{
+		HelloIDs: []ClientHelloID{
+			HelloChrome_Auto,
+			HelloFirefox_Auto,
+			HelloIOS_Auto,
+		},
+		TcpDialTimeout:      time.Second * time.Duration(7+int(tcpDialTimeoutInc)),
+		TlsHandshakeTimeout: time.Second * time.Duration(11+int(tlsHandshakeTimeoutInc)),
+	}
+
+	// choose random clientHelloID to try next
+	nextHelloIdx, err := getRandUint16()
+	if err != nil {
+		return nil, err
+	}
+	r.NextHelloID = r.HelloIDs[int(nextHelloIdx)%len(r.HelloIDs)]
+
+	return r, nil
+}
+
+// Dial attempts to establish connection to given address using different HelloIDs.
+// If a working HelloID is found, it is used again for subsequent Dials.
+// If tcp connection fails or all HelloIDs are tried, returns with last error.
+//
+// Usage examples:
+//    Dial("tcp4", "google.com:443", "google.com")
+//    Dial("tcp", "10.23.144.22:443", "mywebserver.org")
+func (c *Roller) Dial(network, addr, serverName string) (*UConn, error) {
+	remainingHelloIds := make([]ClientHelloID, len(c.HelloIDs))
+	copy(remainingHelloIds, c.HelloIDs)
+
+	c.HelloIDMu.Lock()
+	nextHelloId := c.NextHelloID // keep using same helloID, if it works
+	c.HelloIDMu.Unlock()
+	for {
+		tcpConn, err := net.DialTimeout(network, addr, c.TcpDialTimeout)
+		if err != nil {
+			return nil, err
+		}
+
+		client := UClient(tcpConn, nil, nextHelloId)
+		client.SetSNI(serverName)
+		client.SetDeadline(time.Now().Add(c.TlsHandshakeTimeout))
+		err = client.Handshake()
+		client.SetDeadline(time.Time{}) // unset timeout
+		if err != nil {
+			for i, ID := range remainingHelloIds {
+				if ID == nextHelloId {
+					remainingHelloIds = append(remainingHelloIds[:i], remainingHelloIds[i+1:]...)
+					break
+				}
+			}
+			if len(remainingHelloIds) == 0 {
+				return nil, err
+			}
+			nextHelloIdIdx, err := getRandUint16()
+			if err != nil {
+				return nil, err
+			}
+			nextHelloId = remainingHelloIds[int(nextHelloIdIdx)%len(remainingHelloIds)]
+			continue
+		}
+		c.HelloIDMu.Lock()
+		c.NextHelloID = nextHelloId
+		c.HelloIDMu.Unlock()
+		return client, err
+	}
+}
+
+func getRandUint16() (uint16, error) {
```

Comment:
It seems like all the calls to this function would be better off using [`rand.Intn`](https://golang.org/pkg/math/rand/#Rand.Intn) or [`crypto.rand.Int`](https://golang.org/pkg/crypto/rand/#Int)

Reply:
`rand.Intn` is insecure
`crypto.rand.Int` takes `rand.Reader`, and `big.Int` and returns `big.Int`(!) and `error`, which is a pain to deal with.
I can't see how either of those is better.

Reply:
If you don't like either of those, then I think you should implement an `Intn` style function instead of this one, and include a comment explaining why you didn't use either of the stdlib's options.

Reply:
Just remembered, I already made `getRandInt`, which takes simple `int` as maxValue, uses `big.Int` inside, and returns (`int`, `error`). This is a nice to use function, so I'll switch to that :)